src/Microsoft.AspNetCore.Mvc.Core/Controllers/DefaultControllerTypeProvider.cs

```diff
@@ -44,79 +44,48 @@ public virtual IEnumerable<TypeInfo> ControllerTypes
         /// <returns><c>true</c> if the <paramref name="typeInfo"/> is a controller. Otherwise <c>false</c>.</returns>
-        protected internal virtual bool IsController(
-            TypeInfo typeInfo,
-            ISet<Assembly> candidateAssemblies)
+        protected internal virtual bool IsController(TypeInfo typeInfo)
         {
             if (typeInfo == null)
             {
                 throw new ArgumentNullException(nameof(typeInfo));
             }
 
-            if (candidateAssemblies == null)
-            {
-                throw new ArgumentNullException(nameof(candidateAssemblies));
-            }
-
             if (!typeInfo.IsClass)
             {
                 return false;
             }
+
             if (typeInfo.IsAbstract)
             {
                 return false;
             }
+
             // We only consider public top-level classes as controllers. IsPublic returns false for nested
             // classes, regardless of visibility modifiers
             if (!typeInfo.IsPublic)
             {
                 return false;
             }
+
             if (typeInfo.ContainsGenericParameters)
             {
                 return false;
             }
-            if (!typeInfo.Name.EndsWith(ControllerTypeName, StringComparison.OrdinalIgnoreCase) &&
-                !DerivesFromController(typeInfo, candidateAssemblies))
-            {
-                return false;
-            }
+
             if (typeInfo.IsDefined(typeof(NonControllerAttribute)))
             {
                 return false;
             }
 
-            return true;
-        }
-
-        private bool DerivesFromController(TypeInfo typeInfo, ISet<Assembly> candidateAssemblies)
-        {
-            while (typeInfo != ObjectTypeInfo)
+            if (!typeInfo.Name.EndsWith(ControllerTypeName, StringComparison.OrdinalIgnoreCase) &&
+                !typeInfo.IsDefined(typeof(ControllerAttribute)))
             {
-                var baseTypeInfo = typeInfo.BaseType.GetTypeInfo();
-
-                // A base type will be treated as a controller if
-                // a) it ends in the term "Controller" and
-                // b) it's assembly is one of the candidate assemblies we're considering. This ensures that the assembly
-                // the base type is declared in references Mvc.
-                if (baseTypeInfo.Name.EndsWith(ControllerTypeName, StringComparison.Ordinal) &&
-                    candidateAssemblies.Contains(baseTypeInfo.Assembly))
-                {
-                    return true;
-                }
-
-                // c). The base type is called 'Controller.
-                if (string.Equals(baseTypeInfo.Name, ControllerTypeName, StringComparison.Ordinal))
-                {
-                    return true;
-                }
-
-                typeInfo = baseTypeInfo;
+                return false;
             }
 
-            return false;
+            return true;
```

Comment:
Sooooooo much simpler. The reason for this change is that it's now possible to **explain** the rules.

1. Inherit from a class with `[Controller]` (includes our base-classes)
2. Use the suffix "Controller"

Those are the only rules